ghproxy/ghmetrics/ghmetrics.go

```diff
@@ -0,0 +1,91 @@
+package ghmetrics
+
+import (
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+)
+
+// ghTokenUntilResetGaugeVec provides the 'github_token_reset' gauge that
+// enables keeping track of GitHub reset times.
+var ghTokenUntilResetGaugeVec = prometheus.NewGaugeVec(
+	prometheus.GaugeOpts{
+		Name: "github_token_reset",
+		Help: "Last reported GitHub token reset time.",
+	},
+	[]string{"token_hash"},
+)
+
+// ghTokenUsageGaugeVec provides the 'github_token_usage' gauge that
+// enables keeping track of GitHub calls and quotas.
+var ghTokenUsageGaugeVec = prometheus.NewGaugeVec(
+	prometheus.GaugeOpts{
+		Name: "github_token_usage",
+		Help: "How many GitHub token requets have been used up.",
+	},
+	[]string{"token_hash"},
+)
+
+// ghRequestsGauge provides the 'github_requests' gauge that keeps track
+// of the number of GitHub requests by API path.
+var ghRequestsGauge = prometheus.NewGaugeVec(
+	prometheus.GaugeOpts{
+		Name: "github_requests",
+		Help: "GitHub requests by API path.",
+	},
+	[]string{"token_hash", "path", "status", "duration"},
+)
+
+var muxTokenUsage, muxRequestMetrics sync.Mutex
+var lastGitHubResponse time.Time
+
+func init() {
+	prometheus.MustRegister(ghTokenUntilResetGaugeVec)
+	prometheus.MustRegister(ghTokenUsageGaugeVec)
+	prometheus.MustRegister(ghRequestsGauge)
+}
+
+// CollectGithubTokenMetrics publishes the rate limits of the github api to
+// `github_token_usage` on prometheus.
+func CollectGithubTokenMetrics(tokenHash string, headers http.Header, reqStartTime, responseTime time.Time) {
+	remaining := headers.Get("X-RateLimit-Remaining")
+	timeUntilReset := timestampStringToTime(headers.Get("X-RateLimit-Reset"))
```

Comment:
Is this header specified for v4 requests too? We wouldn't want to conflate the v3 and v4 API rate limits.

Reply:
Quickly testing it on the GH api endpoints gives me these headers on the graphql API as well.

Reply:
This will not work then. We need to distinguish the separate rate limits.